### ISSUE TITLE
[restrict-template-expressions] Missing handling of intersection type

### ISSUE THREAD

**Issue by ulrichb** [typescript-eslint/typescript-eslint#1797]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": "error"
  }
}
```

```TS
type MyString = string & { kind: "MyString" };

const myString = "x" as MyString;

console.log(`myString: ${myString}`); // False positive "Invalid type of template ..."
```

**Expected Result**

No error.

**Actual Result**

"Invalid type of template literal expression"

**Additional Info**

Real-world example:

```ts
import { Term } from "sparqljs";

const someTerm = "x" as Term;

console.log(`someTerm: ${someTerm}`);
```

See https://github.com/DefinitelyTyped/DefinitelyTyped/blob/638a579eecfa0486115f2440518603f25a8b58b7/types/sparqljs/index.d.ts#L223.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.25.0` |
| `@typescript-eslint/parser`        | `2.25.0` |

**Proposed solution**

Don't emit an error if _any_ item of the intersection type is "allowed" in respect to the options.

**Comment by bradzacher** [comment]
Happy to accept a PR!
This wasn't handled because intrinsically, the type is not the primitive type.

**Comment by ulrichb** [comment]
>  This wasn't handled because intrinsically, the type is not the primitive type.

Yes that's correct, but this pattern (add a "fake" disciminator/tag prop) is nice to "type" primitives (forbid assignability between `string` / `MyStringA` / `MyStringB`).

**Comment by bradzacher** [comment]
yeah, I use this within the parser here to help distinguish between known-canonicalized paths, and other paths.

It would be a good change to make, we should probably create some utilities for this, because this will effect many of our rules.
